### DISCUSSION TITLE
Any reason translation registration needs a validator instance?

### DISCUSSION THREAD

**Discussion by RangelReale** [go-playground/validator#1077] [General]
Each translation has a registration function with this signature:

```
func RegisterDefaultTranslations(v *validator.Validate, trans ut.Translator) (err error)
```

It needs a `*validator.Validate` instance because [it calls RegisterTranslation on it](https://github.com/go-playground/validator/blob/8f07b0368280d9fc70b4ffd3708b31427bbc286d/translations/en/en.go#L1375), but I can't see why it couldn't just only register them in `ut.Translator` together with `translateFunc`, and the validator instance would then fetch this list from it. This would enable creating multiple `Validator` instances using the same `Translator`, which is impossible today as registering translations in each instace overrides the Translator instance and errors out on duplicated translation.

For example, we could change the translator registration code to this (current code is commented before the changes):

```go
		if t.customTransFunc != nil && t.customRegisFunc != nil {
			// err = v.RegisterTranslation(t.tag, trans, t.customRegisFunc, t.customTransFunc)
			err = t.customRegisFunc(trans, t.customTransFunc)(trans)
		} else if t.customTransFunc != nil && t.customRegisFunc == nil {
			// err = v.RegisterTranslation(t.tag, trans, registrationFunc(t.tag, t.translation, t.override), t.customTransFunc)
			err = registrationFunc(t.tag, t.translation, t.override, t.customTransFunc)(trans)
		} else if t.customTransFunc == nil && t.customRegisFunc != nil {
			// err = v.RegisterTranslation(t.tag, trans, t.customRegisFunc, translateFunc)
			err = t.customRegisFunc(trans, translateFunc)(trans)
		} else {
			// err = v.RegisterTranslation(t.tag, trans, registrationFunc(t.tag, t.translation, t.override), translateFunc)
			err = registrationFunc(t.tag, t.translation, t.override, translateFunc)(trans)
		}
```

We would need to modify `Translator.Add` to accept and store a translateFunc, and some function to fetch this list. This translateFunc has no hard binding to a Validator instance, so it should not be a problem.

Then we could have a method like `func (v *Validate) RegisterTranslationInstance(trans ut.Translator)` that would just copy the list from `trans.translations` to its internal `v.transTagFunc` to keep the same pattern as today.

**Comment by RangelReale** [comment]
My use case is that I need different rules for the same struct in different parts of my code, so I would need a new `Validate` instance for each function call, but without this I would need to reload and parse all translationsagain every time.

**Reply by zemzale** [reply]
I would like to add here that having multiple `Validate` instances is an anti-pattern since the validator is designed to be a singleton. It does caching to improve performance. 

See more here: 
https://pkg.go.dev/github.com/go-playground/validator/v10#hdr-Singleton

**Reply by RangelReale** [reply]
yes, I decided to stop overthinking problems in the future and just using a singleton, thanks!

**Comment by RangelReale** [comment]
Hmm I see, `universal-translator` is a separate package that is meant to only handle the strings part. Will have to think more about this.

**Comment by RangelReale** [comment]
I created a PR #1079 with a simpler solution to this problem, please see what you think.
